### PR TITLE
TS-4821: Make hwloc a required dependency

### PR DESCRIPTION
**Pull request by PSUdaemon**

### CODE REVIEW

**Comment by atsci** [comment]
FreeBSD build *failed*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/727/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/623/ for details.

**Comment by atsci** [comment]
FreeBSD build *failed*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/730/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/626/ for details.

File: build/hwloc.m4 [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  PKG_CHECK_MODULES([HWLOC], [hwloc], [
+    SAVE_LIBS="$LIBS"
+    LIBS="-lhwloc"
+    AC_LANG_PUSH([C++])
+    AC_MSG_CHECKING([for hwloc C++ linking])
+    AC_LINK_IFELSE([
+      AC_LANG_PROGRAM([#include <hwloc.h>],[hwloc_topology_t t; hwloc_topology_init(&t); hwloc_get_type_depth(t, HWLOC_OBJ_SOCKET);])],[
+      AC_SUBST([HWLOC_CFLAGS])
+      AC_SUBST([HWLOC_LIBS])
+      AC_MSG_RESULT([yes])
```
Comment (jpeach) [inline build/hwloc.m4]: You can simplify the ``AC_LINK_IFELSE`` with ``AC_CHECK_LIB``.

File: build/hwloc.m4 [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+dnl limitations under the License.
+
+dnl
+dnl hwloc.m4: Trafficserver's hwloc autoconf macros
+dnl
+
+dnl
+dnl TS_CHECK_HWLOC: look for hwloc libraries and headers
+dnl
+AC_DEFUN([TS_CHECK_HWLOC], [
```
Comment (jpeach) [inline build/hwloc.m4]: Normally the macro would just do the check, taking arguments of what to do if the check succeeded or failed. The caller would be responsible for throwing the error.

**Comment by jpeach** [comment]
@PSUdaemon don't forget to update the docs when you land this ... there are sections that talk about dependencies and building from source.

**Comment by atsci** [comment]
FreeBSD build *failed*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/882/ for details.

**Comment by atsci** [comment]
Linux build *failed*! See https://ci.trafficserver.apache.org/job/Github-Linux/777/ for details.

**Comment by PSUdaemon** [comment]
@jpeach we should close this, right? You found a case where we can't expect hwloc?

**Comment by PSUdaemon** [comment]
@jpeach can you elaborate on which platform it was that didn't have hwloc? If not, we are going to try to get this in still.

**Comment by PSUdaemon** [comment]
I'll try to rebase this soon.

**Comment by jpeach** [comment]
The problem I saw was with hwloc crashing on startup. This was on a Chinese VM hosting service and was caused by hwloc doing a divide by zero. It was unclear what version of hwloc the bug was fixed in. Needless to say this was a bit of a showstopper for running Traffic Server.

**Comment by PSUdaemon** [comment]
What was the underlying OS?

**Comment by atsci** [comment]
FreeBSD build *failed*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/1271/ for details.

**Comment by atsci** [comment]
Linux build *failed*! See https://ci.trafficserver.apache.org/job/Github-Linux/1166/ for details.

**Comment by jpeach** [comment]
Ubuntu 12 IIRC. But it is a problem with the virtualized hardware not the OS itself.

**Comment by zwoop** [comment]
@PSUdaemon What do you want to do here ? Merge it ?

**Comment by PSUdaemon** [comment]
@zwoop /shrug. I think I need help getting the autoconf stuff to work. It was very circular and hurt my brain.

**Comment by zwoop** [comment]
[approve ci]

**Comment by bryancall** [comment]
@PSUdaemon  Please fix CI issues with this PR.

**Comment by bryancall** [comment]
[approve ci]

**Comment by bryancall** [comment]
[approve ci]

**Comment by zwoop** [comment]
Lets make this happen this week.

**Comment by bryancall** [comment]
@PSUdaemon can we meet today at finish this?

**Comment by bryancall** [comment]
@PSUdaemon 
Do you still want to land this?  If so can you please get it ready to merge.

**Comment by thelounge-zz** [comment]
what is the point of making it required?

in case of virtual machines running on modern hardware where you don't have multiple sockets at all since you get many cores with a single socket it just introduces runtime overhead and additional points of possible bugs

**Comment by bryancall** [comment]
This PR hasn't been updated in more then 2 months and I am closing out PRs that haven't had any updates.  Please feel free to reopen if want to update it.

**Pull request closed without merging**